### PR TITLE
[USMON-1010] unify http2 plaintext and TLS implementations

### PR DESCRIPTION
**Pull request by guyarb**
### What does this PR do?

The PR unifies plaintext and TLS implementations for HTTP2 monitoring.
The PR uses `pktbuf` as an abstraction layer to achieve maximum shared code between the plaintext and TLS versions

The PR removes redundant and duplicated functions.

### Motivation

Remove irrelevant code base and achieve single logic between plain http2 and TLS http2.

#### Complexity changes
```
| Filename/Program                         |   Stack Usage |   Instructions Processed |   Instructions Processed limit |   Max States per Instruction |   Peak States |   Total States |
|------------------------------------------|---------------|--------------------------|--------------------------------|------------------------------|---------------|----------------|
| usm/socket__http2_dynamic_table_cleaner  |            -8 |                       -9 |                              0 |                            0 |            -1 |             -1 |
| usm/socket__http2_eos_parser             |            -8 |                       -1 |                              0 |                            0 |             0 |              0 |
| usm/socket__http2_filter                 |            16 |                   143854 |                              0 |                          -11 |          1753 |           2459 |
| usm/socket__http2_handle_first_frame     |             8 |                        3 |                              0 |                            0 |             4 |              3 |
| usm/socket__http2_headers_parser         |           -40 |                  -115166 |                              0 |                           -3 |           183 |          -3779 |
| usm/uprobe__http2_dynamic_table_cleaner  |             0 |                      -17 |                              0 |                            0 |            -1 |             -1 |
| usm/uprobe__http2_tls_eos_parser         |             8 |                        1 |                              0 |                            0 |             0 |              0 |
| usm/uprobe__http2_tls_filter             |            24 |                   136782 |                              0 |                           17 |          2598 |           4524 |
| usm/uprobe__http2_tls_handle_first_frame |            24 |                     -105 |                              0 |                            0 |            -6 |             -7 |
| usm/uprobe__http2_tls_headers_parser     |            24 |                   124410 |                              0 |                            4 |         -4611 |           3820 |
| usm/uprobe__http2_tls_termination        |             0 |                        0 |                              0 |                            0 |             0 |              0 |
```

##### TLS
No CPU change
image

No memory change 
image

Accuracy improved
image

##### Plaintext

CPU slight reduction

image

Memory slight reduction

Accuracy didn't change

image

### CODE REVIEW

File: pkg/network/ebpf/c/protocols/http2/decoding.h [hunk 1]
```diff
@@ -63,28 +171,52 @@ static __always_inline bool parse_field_literal(struct __sk_buff *skb, skb_info_
...
     }
     headers_to_process->original_index = index;
-    headers_to_process->new_dynamic_value_offset = skb_info->data_off;
+    headers_to_process->new_dynamic_value_offset = pktbuf_data_offset(pkt);
     headers_to_process->new_dynamic_value_size = str_len;
     headers_to_process->is_huffman_encoded = is_huffman_encoded;
     // If the string len (`str_len`) is in the range of [0, HTTP2_MAX_PATH_LEN], and we don't exceed packet boundaries
     // (skb_info->data_off + str_len <= skb_info->data_end) and the index is kIndexPath, then we have a path header,
     // and we're increasing the counter. In any other case, we're not increasing the counter.
     *interesting_headers_counter += (str_len > 0 && str_len <= HTTP2_MAX_PATH_LEN);
```
Comment (Yumasi) [inline pkg/network/ebpf/c/protocols/http2/decoding.h:180]: This comment still references `skb_info`, it needs to be updated. 

Additionally, it seems to explain and list what the code until there does check, and restates what the previous comments in the function already explained. One should be able to understand all of that by just reading the code, so I am not sure what is the added value of this comment.

I suggest listing the conditions in function's comment, and maybe have the inner comments reference those conditions instead, if the code is not clear enough for it to be obvious.

**Review by Yumasi — CHANGES_REQUESTED** [review]
Great to see the pktbuf in action! Not much to change appart from a few outdated comments, and nit on some variable names.

**Comment by guyarb** [comment]
rebased on top of [guy.arbitman/USMON-1010-pktbuf-prep](https://github.com/DataDog/datadog-agent/pull/26242)
removed a commit that introduced a verifier issue

**Comment by guyarb** [comment]
rebased on top of main

File: pkg/network/ebpf/c/protocols/http2/decoding.h [hunk 2]
```diff
@@ -577,26 +669,61 @@ int socket__http2_handle_first_frame(struct __sk_buff *skb) {
...
 
         iteration_value->frames_count = 0;
-        bpf_map_update_elem(&http2_remainder, &dispatcher_args_copy.tup, &new_frame_state, BPF_ANY);
+        bpf_map_update_elem(&http2_remainder, tup, &new_frame_state, BPF_ANY);
         // Not calling the next tail call as we have nothing to process.
-        return 0;
+        return;
     }
     // Overriding the data_off field of the cached skb_info. The next prog will start from the offset of the next valid
     // frame.
```
[the remark this answers is not in the document]
Comment (guyarb) [reply pkg/network/ebpf/c/protocols/http2/decoding.h:672]: done

File: pkg/network/ebpf/c/protocols/http2/decoding.h (hunk 1 above)
Comment (guyarb) [reply pkg/network/ebpf/c/protocols/http2/decoding.h:180]: moved to the function's documentation

**Review by Yumasi — APPROVED** [review]

**Pull request merged**